actors/builtin/miner/miner_state.go

```diff
@@ -360,367 +385,400 @@ func (st *State) ForEachSector(store adt.Store, f func(*SectorOnChainInfo)) erro
 		if err != nil {
-			return xerrors.Errorf("failed to lookup sector expirations at epoch %v: %w", epochSet.epoch, err)
+			return err
 		}
 
-		bf, err = bitfield.MergeBitFields(bf, bitfield.NewFromSet(epochSet.sectors))
-		if err != nil {
-			return xerrors.Errorf("failed to update sector expirations at epoch %v: %w", epochSet.epoch, err)
-		}
-		count, err := bf.Count()
+		partitionsArr, err := dl.PartitionsArray(store)
 		if err != nil {
-			return xerrors.Errorf("failed to count sector expirations at epoch %v: %w", epochSet.epoch, err)
-		}
-		if count > SectorsMax {
-			return fmt.Errorf("too many sectors at expiration %d, %d, max %d", epochSet.epoch, count, SectorsMax)
+			return err
 		}
 
-		if err = arr.Set(uint64(epochSet.epoch), bf); err != nil {
-			return xerrors.Errorf("failed to set sector expirations at epoch %v: %w", epochSet.epoch, err)
+		partitionNumbers := make([]uint64, 0, len(partitions))
+		for partIdx := range partitions {
+			partitionNumbers = append(partitionNumbers, partIdx)
 		}
-	}
+		sort.Slice(partitionNumbers, func(i, j int) bool {
+			return partitionNumbers[i] < partitionNumbers[j]
+		})
 
-	st.SectorExpirations, err = arr.Root()
-	if err != nil {
-		return xerrors.Errorf("failed to persist sector expirations: %w", err)
-	}
-	return nil
-}
+		for _, partIdx := range partitionNumbers {
+			var partition Partition
+			found, err := partitionsArr.Get(partIdx, &partition)
+			if err != nil {
+				return err
+			}
+			if !found {
+				continue
+			}
 
-// Removes some sector numbers from the set expiring at an epoch.
-func (st *State) RemoveSectorExpirations(store adt.Store, sectors ...*SectorOnChainInfo) error {
-	if len(sectors) == 0 {
-		return nil
-	}
+			sectorNos := partitions[partIdx]
+			foundSectors, err := bitfield.IntersectBitField(bitfield.NewFromSet(sectorNos), partition.Sectors)
+			if err != nil {
+				return err
+			}
 
-	arr, err := adt.AsArray(store, st.SectorExpirations)
-	if err != nil {
-		return xerrors.Errorf("failed to load sector expirations: %w", err)
-	}
+			// Anything to update?
+			if empty, err := foundSectors.IsEmpty(); err != nil {
+				return err
+			} else if empty {
+				// no.
+				continue
+			}
 
-	for _, epochSet := range groupSectorsByExpiration(sectors) {
-		bf := new(abi.BitField)
-		_, err = arr.Get(uint64(epochSet.epoch), bf)
-		if err != nil {
-			return xerrors.Errorf("failed to lookup sector expirations at epoch %v: %w", epochSet.epoch, err)
-		}
+			if updated, err := partitionCb(dl, &partition, uint64(dlIdx), partIdx, foundSectors); err != nil {
+				return err
+			} else if updated {
+				empty, err := partition.Sectors.IsEmpty()
+				if err != nil {
+					return err
+				}
 
-		bf, err = bitfield.SubtractBitField(bf, bitfield.NewFromSet(epochSet.sectors))
-		if err != nil {
-			return xerrors.Errorf("failed to update sector expirations at epoch %v: %w", epochSet.epoch, err)
-		}
+				if empty {
+					err = partitionsArr.Delete(partIdx)
```

Comment:
This will never happen given the known call site. But also this is unsafe: the partitions array must be dense, we can never remove items except when compacting and re-indexing them. Just use `Set`.